app/src/main/java/crazydude/com/telemetry/protocol/CrsfProtocol.kt

```diff
@@ -24,125 +24,194 @@ class CrsfProtocol : Protocol {
         private const val GPS_TYPE = 0x02
         private const val ATTITUDE_TYPE = 0x1E
         private const val FLIGHT_MODE = 0x21
+
+        private const val MAX_BUFFER_FILL_LIMIT = 256
+        private const val MIN_BUFFER_FILL_LEVEL_BEFORE_LOOKING_FOR_VALID_PACKETS = 20 //aribitrary number which is probably bigger than most packets
+        private const val MAX_PAYLOAD_SIZE = 62
+        private const val MIN_PAYLOAD_SIZE = 6
+        private const val MIN_FLIGHT_MODE_PACKET_LEN = 6
+        private const val GPS_PACKET_LEN = 16
+        private const val ATTITUDE_PACKET_LEN = 7
+        private const val BATTERY_PACKET_LEN = 8
     }
 
+
     override fun process(data: Int) {
         buffer.add(data)
-        if (buffer[0] == RADIO_ADDRESS) {
-            if (buffer.size > 5) {
-                val frameLength = buffer[1]
-                if (frameLength > 3 && frameLength < buffer.size - 1) {
-                    val payload = buffer.subList(2, frameLength + 1)
-                    val frameCrc = buffer[frameLength + 1]
-                    crC8.reset()
-                    payload.map { it.toByte() }.forEach { crC8.update(it) }
-                    val calculatedCrc = crC8.value.toUByte().toInt()
-                    if (frameCrc == calculatedCrc) {
-                        proccessFrame(payload.map { it.toByte() }.toByteArray())
-                        Log.d("CrsfProtocol", "Good frame $payload")
-                        buffer = ArrayList(buffer.drop(frameLength + 2))
+        if (buffer.size > MAX_BUFFER_FILL_LIMIT) {
+            buffer.removeAt(0)
+        }
+
+        if (buffer.size > MIN_BUFFER_FILL_LEVEL_BEFORE_LOOKING_FOR_VALID_PACKETS) {
+            //Get and process any valid packets in the input buffer, removing data from the input buffer as we go
+            getAndProcessValidPackets()
+        }
+    }
+
+
+    private fun getAndProcessValidPackets() {
+        var startCharPos = 0
+        var pos          = 0
```

Comment:
Can you please reformat your code so it doesn't include more than one white-space before and after any symbol?